src/zope/interface/adapter.py

```diff
@@ -119,6 +167,52 @@ def _createLookup(self):
         for name in self._delegated:
             self.__dict__[name] = getattr(self._v_lookup, name)
 
+    # Hooks for subclasses to define the types of objects used in
+    # our data structures.
+    # These have to be documented in the docstring, instead of local
+    # comments, because autodoc ignores the comment and just writes
+    # "alias of list"
+    _sequenceType = list
+    _leafSequenceType = tuple
+    _mappingType = dict
+    _providedType = dict
+
+    def _addValueToLeaf(self, existing_leaf_sequence, new_item):
+        """
+        Add the value *new_item* to the *existing_leaf_sequence*, which may
+        be ``None``.
+
+        If *existing_leaf_sequence* is not *None*, it will be an instance
+        of `_leafSequenceType`.
+
+        This method returns the new value to be stored. It may mutate the
+        sequence in place if it was not None and the type is mutable, but
+        it must also return it.
+
+        Subclasses that redefine `_leafSequenceType` should override this method.
+
+        .. versionadded:: 5.3.0
+        """
+        if existing_leaf_sequence is None:
+            return (new_item,)
+        return existing_leaf_sequence + (new_item,)
+
+    def _removeValueFromLeaf(self, existing_leaf_sequence, to_remove):
+        """
+        Remove the item *to_remove* from the (non-None, non-empty) *existing_leaf_sequence*
+        and return the mutated sequence.
+
+        If there is more than one item that is equal to *to_remove* they must all be
+        removed.
+
+        May mutate in place or return a new object.
+
+        Subclasses that redefine `_leafSequenceType` should override this method.
+
+        .. versionadded:: 5.3.0
+        """
+        return tuple([v for v in existing_leaf_sequence if v != to_remove])
```

Comment:
Removing the square parentheses should grant better performances because you will not create a temporary `list` instance that will be immediately discarded:
```
return tuple(v for v in existing_leaf_sequence if v != to_remove)
```

Maybe even better is:
```
return tuple(filter(lambda value: value !=  to_remove, existing_leaf_sequence)
```
but I have no benchmarks about this.

As before, having mutable objects here might have been better.

Reply:
> Removing the square parentheses should grant better performances...but I have no benchmarks about this.

Luckily, there are benchmarks about this very thing at the top of the file!

https://github.com/zopefoundation/zope.interface/blob/c28c205c91e73d3ca5e195bb0392d10046e13be0/src/zope/interface/adapter.py#L33-L47

I tested all three earlier when the code migrated off of the `map` idiom.

> As before, having mutable objects here might have been better.

Perhaps, but because there are many existing pickles with this structure, there's going to have to be a migration process. I'll address that in `zope.component.persistentregistery`

Reply:
Wow! I am amazed :)
https://github.com/zopefoundation/zope.interface/blob/c28c205c91e73d3ca5e195bb0392d10046e13be0/src/zope/interface/adapter.py#L33-L35
TIL, thanks!

Reply:
For the curious: `tuple(some_object)` calls [`PySequence_Tuple(some_object)`](https://github.com/python/cpython/blob/2fe408497e6838b6fb761e42e8f8ffc70bd6f3b1/Objects/abstract.c#L1908); if the `some_object` argument is exactly a `list`, we wind up in [`PyList_AsTuple(some_object)`](https://github.com/python/cpython/blob/2fe408497e6838b6fb761e42e8f8ffc70bd6f3b1/Objects/listobject.c#L2485), which in turn passes a direct reference to the list's C-level `PyObject*` array back to [`_PyTuple_FromArray`](https://github.com/python/cpython/blob/2fe408497e6838b6fb761e42e8f8ffc70bd6f3b1/Objects/tupleobject.c#L414). That function just does some C pointer walking to initialize the new new tuple's C array.

Every other path through `PySequence_Tuple` (except for the trivial case of `tuple(other_tuple)` resort to the much more expensive case of using generic `iter(some_object)` and `next()` calls to populate the tuple.